File: solutions/090.Subsets_II/AC_dfs_2^n.py
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dfs_2^n.py
# Create Date: 2015-02-08 10:22:37
# Descripton:   


from itertools import groupby
class Solution:
    def dfs(self, dep, newS):
        if dep == len(newS):
            self.res.append(self.cur[:])
        else:
            self.dfs(dep + 1, newS)
            for i in newS[dep]:
                self.cur.append(i)
                self.dfs(dep + 1, newS)
            for i in newS[dep]:
                self.cur.pop()

    # @param S, a list of integer
    # @return a list of lists of integer
    def subsetsWithDup(self, S):
        S.sort()
        newS = [list(g) for k, g in groupby(S)]
        self.res = []
        self.cur = []
        self.dfs(0, newS)
        return self.res

# debug
s = Solution()
print s.subsetsWithDup([1, 2, 2])

File: solutions/090.Subsets_II/AC_dfs_2^n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dfs_2^n.java
*  Create Date: 2015-02-08 11:18:01
*  Descripton:   
*/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Solution {

    public List<List<Integer>> subsetsWithDup(int[] S) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> cur = new ArrayList<Integer>();

        Arrays.sort(S);

        dfs(0, res, cur, S);
        return res;
    }

    private void dfs(int dep, List<List<Integer>> res, List<Integer> cur,
                int[] S) {
        if (dep == S.length) {
            res.add(new ArrayList<Integer>(cur));
        } else {
            int upper = dep;
            while (upper >= 0 && upper < S.length - 1 && S[upper] == S[upper + 1]) {
                ++upper;
            }
            // no choose
            dfs(upper + 1, res, cur, S);
            // choose
            for (int i = dep; i <= upper; ++i) {
                cur.add(new Integer(S[dep]));
                dfs(upper + 1, res, cur, S);
            }
            for (int i = dep; i <= upper; ++i)
                cur.remove(cur.size() - 1);
        }
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {1, 2, 2};
        List<List<Integer>> res = s.subsetsWithDup(input);
        for (List<Integer> i : res) {
            for (Integer j : i)
                System.out.print(j.toString() + ' ');
            System.out.println("");
        }
    }
}
